### PR TITLE
[Qwen] avoid creating attention masks when there is no padding

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

This pull request sets all ones masks to None.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by kashif** [comment]
will do!

File: src/diffusers/pipelines/qwenimage/pipeline_qwenimage_img2img.py [hunk 1]
```diff
@@ -217,14 +218,7 @@ def _get_qwen_prompt_embeds(
...
         split_hidden_states = [e[drop_idx:] for e in split_hidden_states]
-        attn_mask_list = [torch.ones(e.size(0), dtype=torch.long, device=e.device) for e in split_hidden_states]
-        max_seq_len = max([e.size(0) for e in split_hidden_states])
-        prompt_embeds = torch.stack(
-            [torch.cat([u, u.new_zeros(max_seq_len - u.size(0), u.size(1))]) for u in split_hidden_states]
-        )
-        encoder_attention_mask = torch.stack(
-            [torch.cat([u, u.new_zeros(max_seq_len - u.size(0))]) for u in attn_mask_list]
-        )
+        prompt_embeds, encoder_attention_mask = build_prompt_embeds_and_mask(split_hidden_states)
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/qwenimage/pipeline_qwenimage_img2img.py]: curiously, cant we just do

```py
if encoder_attention_mask.all():
    encoder_attention_mask = None
```

Comment (kashif) [reply src/diffusers/pipelines/qwenimage/pipeline_qwenimage_img2img.py]: the issue was that this was giving trouble with torch.compile...

Comment (yiyixuxu) [reply src/diffusers/pipelines/qwenimage/pipeline_qwenimage_img2img.py]: ohh, but I think we typically just torch.compile the models (e.g. `pipeline.transformer.compile(...)`, `pipeline.text_encoder.compile()`)
the computations on the pipeline won't be part of it though

Comment (kashif) [reply src/diffusers/pipelines/qwenimage/pipeline_qwenimage_img2img.py]: oh ok let me check that then

Comment (kashif) [reply src/diffusers/pipelines/qwenimage/pipeline_qwenimage_img2img.py]: added checks for `.all()`

File: tests/models/transformers/test_models_transformer_qwenimage.py [hunk 2]
```diff
@@ -276,3 +276,45 @@ def prepare_dummy_input(self, height, width):
...
     def test_torch_compile_recompilation_and_graph_break(self):
         super().test_torch_compile_recompilation_and_graph_break()
+
+    def test_torch_compile_with_and_without_mask(self):
+        """Test that torch.compile works with both None mask and padding mask."""
+        init_dict, inputs = self.prepare_init_args_and_inputs_for_common()
+        model = self.model_class(**init_dict).to(torch_device)
+        model.eval()
+
+        compiled_model = torch.compile(model, mode="default", fullgraph=False)
```
Comment (sayakpaul) [inline tests/models/transformers/test_models_transformer_qwenimage.py]: Some notes:

* Usually, it should be `model.compile()` as it doesn't wrap the underlying `model` into a dynamo wrapper. This way, we don't have to add any extra code to handle it.
* Why is `fullgraph=False` here?

File: tests/models/transformers/test_models_transformer_qwenimage.py [hunk 3]
```diff
@@ -276,3 +276,45 @@ def prepare_dummy_input(self, height, width):
...
+
+        # Test 1: Run with None mask (no padding, all tokens are valid)
+        inputs_no_mask = inputs.copy()
+        inputs_no_mask["encoder_hidden_states_mask"] = None
+
+        with torch.no_grad():
+            output_no_mask = compiled_model(**inputs_no_mask)
```
Comment (sayakpaul) [inline tests/models/transformers/test_models_transformer_qwenimage.py]: Does it lead to graph breaks? If not, then we should add additional contexts:
https://github.com/huggingface/diffusers/blob/3996788b602eaae4da41a1d45726b62e662b73cf/tests/models/test_modeling_common.py#L2145

**Comment by dxqb** [comment]
does this PR lack a merge, or is the amount of code changes intentional and really part of this PR only? (+318 - 170)
if intentional, could you explain what it does and why all these changes are necessary?

avoiding masks if they're not necessary has been one line before
`attention_mask = attention_mask if not torch.all(text_attention_mask) else None`
(on a boolean mask)

**Comment by kashif** [comment]
@dxqb i intially mis-calculated and was adding support for the pipeline to be compiled, i will revert and simplify

**Comment by dxqb** [comment]
> @dxqb i intially mis-calculated and was adding support for the pipeline to be compiled, i will revert and simplify

thanks!
regarding compile,
 - [Regional compilation](https://huggingface.co/docs/diffusers/optimization/fp16#regional-compilation) is usually as efficient as full compilation (even if it was possible). Regional compilation compiles the transformer blocks, but not the entire pipeline.
 - whenever you want to branch depending on GPU data, that's a graph break for torch.compile. Either inefficient, or fails depending on `fullgraph`. Or, in less abstract terms:
 you want to set your attention mask to `None` if the entire attention mask is `True`. But the attention mask lives on GPU. Checking whether all values of a tensor on GPU are `True` requires a transfer back to CPU - that's always a graph break and cannot be compiled (efficiently).

Therefore, I'd suggest to
 - check for an all-True mask in the pipeline
 - pass a None-Mask to the transformer block in that case - but don't check any GPU data in the transformer block, so the transformer block can be compiled.

**Comment by dxqb** [comment]
> Thanks!
> 
> Could you update with the following things?
> 
>     * Shed light into what caused the speed regression
> 

Here is a benchmark of the impact of using a mask unnecessarily (second graph): https://github.com/huggingface/diffusers/pull/12870#issue-3751633383
torch SDPA falls back to a non-flash algorithm if a mask is used.

**Comment by kashif** [comment]
yes i also benchmarked the mask and got:

nsight_sdpa_mask_regression

**Comment by kashif** [comment]
thanks @dxqb please check now

**Comment by DefTruth** [comment]
I tested this https://github.com/huggingface/diffusers/pull/12987 on L20, and there was an 11% performance improvement in my test.

**Comment by sayakpaul** [comment]
And do you find that to resolve your concerns?

**Comment by DefTruth** [comment]
I think this has solved my problem. Based on this PR, the performance in my test cases has been working as expected.

**Comment by dxqb** [comment]
> thanks @dxqb please check now

I don't have a simple test case, but code looks simple and good now

**Comment by yiyixuxu** [comment]
thanks again @kashif !!
